From 9d3d9392a1b2b5b18b0d5712b54d8893ac01d064 Mon Sep 17 00:00:00 2001
From: Alexey Makhalov <alexey.makhalov@broadcom.com>
Date: Sun, 24 Aug 2025 18:28:03 +0000
Subject: [PATCH 02/02] crypto/algapi.c: honor fcw_skip_tests()

linux-esx does not need to run alg tests in non fips mode. Skip triggerring
them from algapi layer.

algboss.c within the canister boundary already handles fcw_skip_tests(),
without corresponding change in algapi.c we catch WARN_ON:

[    0.913791] WARNING: CPU: 0 PID: 1 at crypto/api.c:176 crypto_schedule_test+0x66/0x80
[    0.914682] Modules linked in:
[    0.916138] CPU: 0 UID: 0 PID: 1 Comm: swapper/0 Not tainted 6.12.1-20.ph5+ #20
[    0.918684] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.16.3-0-ga6ed6b701f0a-prebuilt.qemu.org 04/01/2014
[    0.922682] RIP: 0010:crypto_schedule_test+0x66/0x80
[    0.924775] Code: c6 3a 82 81 82 bf 01 00 00 00 e8 65 52 a6 ff 48 89 da be 01 00 00 00 48 c7 c7 20 43 25 83 e8 e1 36 9d ff 3d 01 80 00 00 74 ca <0f> 0b 48 8b 5d f8 c9 c3 cc cc cc cc 66 66 2e 0f 1f 84 00 00 00 00
[    0.926682] RSP: 0000:ffffc90000013d38 EFLAGS: 00010297
[    0.929180] RAX: 0000000000000000 RBX: ffffffff8325d808 RCX: c0000000ffffdfff
[    0.930685] RDX: 0000000000000000 RSI: 00000000ffffdfff RDI: ffffffff83254320
[    0.933559] RBP: ffffc90000013d40 R08: 0000000000000000 R09: 0000000000000003
[    0.934681] R10: ffffc90000013af8 R11: ffffffff8314b888 R12: 0000000000000000
[    0.937631] R13: ffffc90000013d58 R14: ffffffff8366f5d0 R15: 0000000000000004
[    0.938686] FS:  0000000000000000(0000) GS:ffff88807dc00000(0000) knlGS:0000000000000000
[    0.942238] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    0.942681] CR2: ffff888003c01000 CR3: 0000000003032001 CR4: 0000000000770ef0
[    0.945860] PKRU: 55555554
[    0.946683] Call Trace:
[    0.947677]  <TASK>
[    0.948603]  ? show_regs+0x64/0x70
[    0.950092]  ? __warn+0x8d/0x140
[    0.950682]  ? crypto_schedule_test+0x66/0x80
[    0.952572]  ? report_bug+0x171/0x1a0
[    0.954093]  ? handle_bug+0x60/0xb0
[    0.954685]  ? exc_invalid_op+0x18/0x70
[    0.956304]  ? asm_exc_invalid_op+0x1b/0x20
[    0.958083]  ? crypto_schedule_test+0x66/0x80
[    0.958681]  ? crypto_schedule_test+0x5f/0x80
[    0.960514]  crypto_register_alg+0x9a/0xe0
[    0.962294]  crypto_register_shashes+0x93/0x170
[    0.962682]  sha3_generic_mod_init+0x1d/0x70
[    0.964422]  fcw_arch_initcall+0x1e/0x50
[    0.966041]  ? __cfi_fcw_arch_initcall+0x10/0x10
[    0.966684]  do_one_initcall+0x59/0x330
[    0.968395]  kernel_init_freeable+0x435/0x645
[    0.970162]  ? __cfi_kernel_init+0x10/0x10
[    0.970681]  kernel_init+0x1a/0x1d0
[    0.972079]  ret_from_fork+0x4a/0x70
[    0.973591]  ? __cfi_kernel_init+0x10/0x10
[    0.974682]  ret_from_fork_asm+0x1a/0x30
[    0.976260]  </TASK>

Signed-off-by: Alexey Makhalov <alexey.makhalov@broadcom.com>
---
 crypto/algapi.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/crypto/algapi.c b/crypto/algapi.c
index 004d27e41..3700c4c5d 100644
--- a/crypto/algapi.c
+++ b/crypto/algapi.c
@@ -20,6 +20,7 @@
 #include <linux/workqueue.h>
 
 #include "internal.h"
+#include "fips_canister_wrapper.h"
 
 static LIST_HEAD(crypto_template_list);
 
@@ -268,6 +269,7 @@ static struct crypto_larval *crypto_alloc_test_larval(struct crypto_alg *alg)
 
 	if (!IS_ENABLED(CONFIG_CRYPTO_MANAGER) ||
 	    IS_ENABLED(CONFIG_CRYPTO_MANAGER_DISABLE_TESTS) ||
+	    fcw_skip_tests() ||
 	    (alg->cra_flags & CRYPTO_ALG_INTERNAL))
 		return NULL; /* No self-test needed */
 
@@ -1022,6 +1024,9 @@ static void __init crypto_start_tests(void)
 	if (IS_ENABLED(CONFIG_CRYPTO_MANAGER_DISABLE_TESTS))
 		return;
 
+	if (fcw_skip_tests())
+		return;
+
 	set_crypto_boot_test_finished();
 
 	for (;;) {
-- 
2.40.4

